src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIMarsHandle.cs

```diff
@@ -0,0 +1,521 @@
+        /// </summary>
+        private void SendAckIfNecessary()
+        {
+            uint receiveHighwater;
+            uint receiveHighwaterLastAck;
+
+            lock (this)
+            {
+                receiveHighwater = _receiveHighwater;
+                receiveHighwaterLastAck = _receiveHighwaterLastAck;
+            }
+
+            if (receiveHighwater - receiveHighwaterLastAck > ACK_THRESHOLD)
+            {
+                SendControlPacket(SNISMUXFlags.SMUX_ACK, true);
+            }
+        }
+
+        /// <summary>
+        /// Receive a packet synchronously
+        /// </summary>
+        /// <param name="packet">SNI packet</param>
+        /// <param name="timeout">Timeout</param>
+        /// <returns>SNI error code</returns>
+        public override uint Receive(ref SNIPacket packet, int timeout)
+        {
+            int queueCount;
+            uint result = TdsEnums.SNI_SUCCESS_IO_PENDING;
+
+            while (true)
+            {
+                lock (_receivedPacketQueue)
+                {
+                    if (_connectionError != null)
+                    {
+                        SNILoadHandle.SingletonInstance.LastError = _connectionError;
+                        return TdsEnums.SNI_ERROR;
+                    }
+
+                    queueCount = _receivedPacketQueue.Count;
+
+                    if (queueCount > 0)
+                    {
+                        packet = _receivedPacketQueue.Dequeue();
+
+                        if (queueCount == 1)
+                        {
+                            _packetEvent.Reset();
+                        }
+
+                        result = TdsEnums.SNI_SUCCESS;
+                    }
+                }
+
+                if (result == TdsEnums.SNI_SUCCESS)
+                {
+                    lock (this)
+                    {
+                        _receiveHighwater++;
+                    }
+
+                    SendAckIfNecessary();
+                    return result;
+                }
+
+                if (!_packetEvent.WaitOne(timeout))
+                {
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, 0, "Timeout error");
+                    return TdsEnums.SNI_WAIT_TIMEOUT;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Check SNI handle connection
+        /// </summary>
+        /// <param name="handle"></param>
+        /// <returns>SNI error status</returns>
+        public override uint CheckConnection()
+        {
+            return _connection.CheckConnection();
+        }
+
+        /// <summary>
+        /// Set async callbacks
+        /// </summary>
+        /// <param name="receiveCallback">Receive callback</param>
+        /// <param name="sendCallback">Send callback</param>
+        public override void SetAsyncCallbacks(SNIAsyncCallback receiveCallback, SNIAsyncCallback sendCallback)
+        {
+        }
```

Comment:
Mars connections have their own callbacks. See constructor for SNIMarsConnection.